src/syntax/ast.rs

```diff
@@ -5,101 +5,122 @@ pub struct Resource {
 }
 
 #[derive(Debug, PartialEq)]
-pub struct Comment {
-    pub body: String,
-}
-
-#[derive(Debug, PartialEq)]
-pub struct Section {
-    pub key: Key,
-    pub body: Vec<Entry>,
+pub struct Message {
+    pub id: Identifier,
+    pub value: Option<Pattern>,
+    pub attributes: Option<Vec<Attribute>>,
     pub comment: Option<Comment>,
 }
 
 #[derive(Debug, PartialEq)]
 pub enum Entry {
     Message(Message),
-    Comment(Comment),
     Section(Section),
+    Comment(Comment),
     Junk(JunkEntry),
 }
 
 #[derive(Debug, PartialEq)]
-pub struct Message {
+pub struct Pattern {
+    pub elements: Vec<PatternElement>,
+    pub quoted: bool,
+}
+
+#[derive(Debug, PartialEq)]
+pub enum Expression {
+    Selector(SelectorExpression),
+    SelectExpression {
+        exp: Option<SelectorExpression>,
+        variants: Vec<Variant>
+    },
+}
+
+#[derive(Debug, PartialEq)]
+pub enum SelectorExpression {
+    Pattern(Pattern),
+    MessageReference(String),
+    ExternalArgument(String),
+    CallExpression {
+        callee: Builtin,
+        args: Vec<SelectorExpression>,
```

Comment:
Shouldn't this just be `Vec<Expression>`? I understand that we don't want to allow bare `SelectExpressions` as arguments here, and they require to be wrapped in `{ }`, but I think this should be encoded in the EBNF rather than the AST. There's no ambiguity in the AST and the braces don't have to be part of it.

OTOH, I see how this helps serialization. If the AST contains the information about the braces used for grouping, the process of serialization doesn't have to check the parent of a `SelectExpression` to decide if it should be wrapped in braces.